libs/hooks/debounce/src/lib/utils/fixed-size-expiring-cache.ts

```diff
@@ -0,0 +1,94 @@
+/**
+ * Wrapper object so we can easily handle falsy/undefined values.
+ */
+type Entry<T> = {
+  value: T;
+  expiry: number;
+};
+
+type Options = {
+  /**
+   * A positive integer setting the max number of items in the cache before the oldest entry is removed.
+   */
+  maxItems: number;
+  /**
+   * Time to live for items in cache in milliseconds.
+   */
+  ttlMs: number;
+};
+
+/**
+ * A very simple cache which lazily evicts and expires keys.
+ * The cache has a fixed max size, and when that size is exceeded, the oldest key is evicted based on insertion order.
+ * When a key is retrieved, if it has expired, it is removed from the cache and undefined is returned.
+ * If a key is set that already exists, it is updated and its recency (but not it's TTL) is updated.
+ */
+export class FixedSizeExpiringCache<T> {
+  private cacheMap = new Map<string, Entry<T>>();
+  private readonly maxItems: number;
+  private readonly ttl: number;
+
+  constructor(options: Options) {
+    if (options.maxItems < 1) {
+      throw new Error('maxItems must be a positive integer');
+    }
+    this.maxItems = options.maxItems;
+    if (options.ttlMs < 1) {
+      throw new Error('ttlMs must be a positive integer');
+    }
+    this.ttl = options.ttlMs;
+  }
+
+  /**
+   * Gets a key from the cache, updating its recency.
+   *
+   * @param key key for the entry
+   * @returns value or key or undefined
+   */
+  get(key: string): T | void {
+    if (key) {
+      const entry = this.cacheMap.get(key);
+      if (entry) {
```

Comment:
Could a valid value ever be an empty string? I wonder if it would be slightly safer to check for undefined.